py3status/modules/mpris.py

```diff
@@ -0,0 +1,385 @@
+        """
+        Get the current metadatas
+        """
+        is_stream = False
+        player = UNKNOWN
+        state = UNKNOWN
+        artist = UNKNOWN
+        album = UNKNOWN
+        length = UNKNOWN_TIME
+        time = UNKNOWN_TIME
+        title = UNKNOWN
+
+        try:
+            player = self._player.Identity
+            (color, state) = self._get_state_format(i3s_config)
+
+            _time_ms = self._player.Position
+            time = _get_time_str(_time_ms)
+
+            metadata = self._player.Metadata
+            if len(metadata) > 0:
+                is_stream = 'file://' not in metadata.get('xesam:url')
+                title = metadata.get('xesam:title') or UNKNOWN
+                album = metadata.get('xesam:album') or UNKNOWN
+
+                if metadata.get('xesam:artist') is not None:
+                    artist = metadata.get('xesam:artist')[0]
+                else:
+                    # we assume here that we playing a video and these types of
+                    # media we handle just like streams
+                    is_stream = True
+
+                _length_ms = metadata.get('mpris:length')
+                if _length_ms is not None:
+                    length = _get_time_str(_length_ms)
+            else:
+                # use stream format if no metadata is available
+                is_stream = True
+        except Exception:
+            color = i3s_config['color_bad']
+
+        if is_stream:
+            # delete the file extension
+            title = re.sub(r'\....$', '', title)
+            format = self.format_stream
+        else:
+            format = self.format
+
+        return (format.format(player=player,
+                              state=state,
+                              album=album,
+                              artist=artist,
+                              length=length,
+                              time=time,
+                              title=title), color)
+
+    def _get_control_states(self):
+        control_states = {
+            'pause':    _get_ctr_struct(self._player.CanPause,
+                                        self.icon_pause,
+                                        self._player.Pause),
+            'play':     _get_ctr_struct(self._player.CanPlay,
+                                        self.icon_play,
+                                        self._player.Play),
+            'stop':     _get_ctr_struct(True,
+                                        self.icon_stop,
+                                        self._player.Stop),
+            'next':     _get_ctr_struct(self._player.CanGoNext,
+                                        self.icon_next,
+                                        self._player.Next),
+            'previous': _get_ctr_struct(self._player.CanGoPrevious,
+                                        self.icon_previous,
+                                        self._player.Previous)
+        }
+
```

Comment:
using `_get_ctr_struct()` makes this hard to read why not just be explicit to start with.

 ``` 
'play': {'clickable': ...,
         'icon': ...,
         'action':...},
```

Also I'd just use the attributes eg `'CanGoPrevious'` and then do `getattr(self._player, attribute)` as needed rather than having to keep generating the dict all the time.